l1-contracts/contracts/bridge/L1Nullifier.sol

```diff
@@ -0,0 +1,757 @@
+                    _transferData: _assetData
+                });
+                if (dataHash != txDataHash) {
+                    revert DepositDoesNotExist();
+                }
+            }
+        }
+        delete depositHappened[_chainId][_l2TxHash];
+    }
+
+    /// @notice Internal function that handles the logic for finalizing withdrawals, supporting both the current bridge system and the legacy ERC20 bridge.
+    /// @param _finalizeWithdrawalParams The structure that holds all necessary data to finalize withdrawal
+    /// @return l1Receiver The address to receive bridged assets.
+    /// @return assetId The bridged asset ID.
+    /// @return amount The amount of asset bridged.
+    function _finalizeWithdrawal(
+        FinalizeWithdrawalParams calldata _finalizeWithdrawalParams
+    ) internal nonReentrant whenNotPaused returns (address l1Receiver, bytes32 assetId, uint256 amount) {
+        bytes memory transferData;
+        (assetId, transferData) = _verifyAndGetWithdrawalData(_finalizeWithdrawalParams);
+
+        (l1Receiver, amount) = l1AssetRouter.finalizeWithdrawal(
+            _finalizeWithdrawalParams.chainId,
+            assetId,
+            transferData
+        );
+    }
+
+    /// @notice Internal function that handles the logic for finalizing withdrawals, supporting both the current bridge system and the legacy ERC20 bridge.
+    /// @param _finalizeWithdrawalParams The structure that holds all necessary data to finalize withdrawal
+    /// @return assetId The bridged asset ID.
+    /// @return transferData The encoded transfer data.
+    function _verifyAndGetWithdrawalData(
+        FinalizeWithdrawalParams calldata _finalizeWithdrawalParams
+    ) internal whenNotPaused returns (bytes32 assetId, bytes memory transferData) {
+        if (
+            isWithdrawalFinalized[_finalizeWithdrawalParams.chainId][_finalizeWithdrawalParams.l2BatchNumber][
+                _finalizeWithdrawalParams.l2MessageIndex
+            ]
+        ) {
+            revert WithdrawalAlreadyFinalized();
+        }
+        isWithdrawalFinalized[_finalizeWithdrawalParams.chainId][_finalizeWithdrawalParams.l2BatchNumber][
+            _finalizeWithdrawalParams.l2MessageIndex
+        ] = true;
+
+        // Handling special case for withdrawal from ZKsync Era initiated before Shared Bridge.
+        require(
+            !_isEraLegacyEthWithdrawal(_finalizeWithdrawalParams.chainId, _finalizeWithdrawalParams.l2BatchNumber),
+            "L1N: legacy eth withdrawal"
+        );
+        require(
+            !_isEraLegacyTokenWithdrawal(_finalizeWithdrawalParams.chainId, _finalizeWithdrawalParams.l2BatchNumber),
+            "L1N: legacy token withdrawal"
+        );
+
+        (assetId, transferData) = _checkWithdrawal(_finalizeWithdrawalParams);
+    }
+
+    function finalizeWithdrawalExternal(FinalizeWithdrawalParams calldata _finalizeWithdrawalParams) external {
```

Comment:
why do we need this method? i.e. why it can not be an internal one

Reply:
because of memory vs calldata for the _finalizeWithdrawalParams struct.